core/lib/Cardano/Address/Derivation.hs

```diff
@@ -333,29 +341,55 @@ instance Bounded (Index 'WholeDomain depth) where
     minBound = Index minBound
     maxBound = Index maxBound
 
-instance Enum (Index 'Hardened depth) where
-    fromEnum (Index ix) = fromIntegral ix
-    toEnum ix
-        | Index (fromIntegral ix) < minBound @(Index 'Hardened _) =
-            error "Index@Hardened.toEnum: bad argument"
-        | otherwise =
-            Index (fromIntegral ix)
-
-instance Enum (Index 'Soft depth) where
-    fromEnum (Index ix) = fromIntegral ix
-    toEnum ix
-        | Index (fromIntegral ix) > maxBound @(Index 'Soft _) =
-            error "Index@Soft.toEnum: bad argument"
-        | otherwise =
-            Index (fromIntegral ix)
-
-instance Enum (Index 'WholeDomain depth) where
-    fromEnum (Index ix) = fromIntegral ix
-    toEnum ix
-        | Index (fromIntegral ix) > maxBound @(Index 'WholeDomain _) =
-            error "Index@WholeDomain.toEnum: bad argument"
-        | otherwise =
-            Index (fromIntegral ix)
+-- | Serialization of 'Index' to and from 'Word32'.
+--
+-- @since 3.3.0
+class Indexed a where
+    fromWord32 :: Word32 -> Maybe a
+    toWord32 :: a -> Word32
```

Comment:
Hi @rvl!

This is just a suggestion, but perhaps it would be better if the class name `Indexed` and function names `{fromWord32,toWord32}` could more clearly relate to one another.

Some thoughts:

- This class deals exclusively with things that are convertible to and from `Word32`.
- It's also a replacement for `Enum` that works with `Word32` instead of `Int`.

Perhaps call it `Enum32`?

Then we'd have:

```hs
class Enum32 a where
    toEnum32 :: Word32 -> Maybe a
    fromEnum32 :: a -> Word32
```
(Note that the `to` and `from` functions are going in the same direction as the `Enum` class.)

Reply:
Thanks for the ideas. I would like to distance this class from `Enum`, which has features and characteristics that we don't need for address path indices. I was thinking SerialiseAsWord32. But actually, now I think this class should be distilled down to just the function `indexFromWord32` (more like what Luite originally did...).